base/setup/lib/utils/partlist.c

```diff
@@ -513,45 +508,117 @@ IsSuperFloppy(
     }
 
-    /* The partition lengths should agree */
-    PartitionLengthEstimate = GetDiskSizeInBytes(DiskEntry);
-    if (PartitionInfo->PartitionLength.QuadPart != PartitionLengthEstimate)
+    /* The partition and disk sizes should agree */
+    if (DiskSize && (PartitionInfo->PartitionLength.QuadPart != *DiskSize))
     {
-        DPRINT1("PartitionLength = %I64u is different from PartitionLengthEstimate = %I64u\n",
-                PartitionInfo->PartitionLength.QuadPart, PartitionLengthEstimate);
+        DPRINT1("PartitionLength = %I64u is different from DiskSize = %I64u\n",
+                PartitionInfo->PartitionLength.QuadPart, *DiskSize);
     }
 
     return TRUE;
 }
 
+BOOLEAN
+IsDiskSuperFloppy(
+    _In_ const DRIVE_LAYOUT_INFORMATION* Layout,
+    _In_opt_ const ULONGLONG* DiskSize)
+{
+    DISK_PARTITION_INFO DiskInfo;
+
+    /* We must have only one partition */
+    if (Layout->PartitionCount != 1)
+        return FALSE;
+
+    /* Build the disk partition info */
+    DiskInfo.SizeOfPartitionInfo = RTL_SIZEOF_THROUGH_FIELD(DISK_PARTITION_INFO, Mbr);
+    DiskInfo.PartitionStyle = PARTITION_STYLE_MBR;
+    DiskInfo.Mbr.Signature = Layout->Signature;
+    DiskInfo.Mbr.CheckSum = 0; // Dummy value
+
+    /* Call the helper on the single partition entry */
+    return IsDiskSuperFloppy2(&DiskInfo, DiskSize, Layout->PartitionEntry);
+}
+
+BOOLEAN
+IsDiskSuperFloppyEx(
+    _In_ const DRIVE_LAYOUT_INFORMATION_EX* LayoutEx,
+    _In_opt_ const ULONGLONG* DiskSize)
+{
+    DISK_PARTITION_INFO DiskInfo;
+    const PARTITION_INFORMATION_EX* PartitionInfoEx;
+    PARTITION_INFORMATION PartitionInfo;
+
+    /* The layout must be MBR and have only one partition */
+    if (LayoutEx->PartitionStyle != PARTITION_STYLE_MBR)
+        return FALSE;
+    if (LayoutEx->PartitionCount != 1)
+        return FALSE;
+
+    /* Build the disk partition info */
+    DiskInfo.SizeOfPartitionInfo = RTL_SIZEOF_THROUGH_FIELD(DISK_PARTITION_INFO, Mbr);
+    DiskInfo.PartitionStyle = PARTITION_STYLE_MBR; // LayoutEx->PartitionStyle;
+    DiskInfo.Mbr.Signature = LayoutEx->Mbr.Signature;
+    DiskInfo.Mbr.CheckSum = 0; // Dummy value
+
+    /* Convert the single partition entry */
+    PartitionInfoEx = LayoutEx->PartitionEntry;
+
+    PartitionInfo.StartingOffset = PartitionInfoEx->StartingOffset;
+    PartitionInfo.PartitionLength = PartitionInfoEx->PartitionLength;
+    PartitionInfo.HiddenSectors = PartitionInfoEx->Mbr.HiddenSectors;
+    PartitionInfo.PartitionNumber = PartitionInfoEx->PartitionNumber;
+    PartitionInfo.PartitionType = PartitionInfoEx->Mbr.PartitionType;
+    PartitionInfo.BootIndicator = PartitionInfoEx->Mbr.BootIndicator;
+    PartitionInfo.RecognizedPartition = PartitionInfoEx->Mbr.RecognizedPartition;
+    PartitionInfo.RewritePartition = PartitionInfoEx->RewritePartition;
+
+    /* Call the helper on the single partition entry */
+    return IsDiskSuperFloppy2(&DiskInfo, DiskSize, &PartitionInfo);
+}
+
+// static
```

Comment:
Can it be made static? Then let's do it. If not, it might be best to strip the comment.
```suggestion
static
```